sphinx/extension.py

```diff
@@ -51,8 +51,13 @@ def verify_needs_extensions(app: Sphinx, config: Config) -> None:
     for extname, reqversion in config.needs_extensions.items():
         extension = app.extensions.get(extname)
         if extension is None:
-            logger.warning(__('The %s extension is required by needs_extensions settings, '
-                              'but it is not loaded.'), extname)
+            logger.warning(
+                __(
+                    'The %s extension is required by needs_extensions settings, '
+                    'but it is not loaded.'
+                ),
+                extname,
+            )
```

Comment:
Maybe a manual tweak here: would it look better to extract a `msg` variable for the localized warning message?

Reply:
Possible I missed something, but this works for me with `ruff` locally:

```diff
-            logger.warning(
-                __(
-                    'The %s extension is required by needs_extensions settings, '
-                    'but it is not loaded.'
-                ),
-                extname,
-            )
+            msg = __('The %s extension is required by needs_extensions settings, '
+                     'but it is not loaded.')
+            logger.warning(msg)
```

Reply:
ah, yes - the `extname`.  That's important...

Reply:
(even with `extname` added back in, it seems to pass `ruff` checks OK.  I think it makes the overall diff marginally easier to comprehend, too, which is sometimes nice)

Reply:
Have you actually removed `sphinx/extension.py` from the ruff format exclude list? because it does not stay like you wrote it, you'll get:

```python
msg = __(
                'The %s extension is required by needs_extensions settings, '
                + 'but it is not loaded.'
            )
```

Reply:
> Have you removed `sphinx/extension.py` from the ruff format exclude list? because it does not stay like you wrote it, you'll get:

Yep, I checked out this pull request branch before applying the change, to make sure it's removed and applies cleanly.

> ```python
>             msg = __(
>                 'The %s extension is required by needs_extensions settings, '
>                 + 'but it is not loaded.'
>             )
> ```

Strange - no reformatting occurs for me.  I'm using `ruff 0.3.2`.

The improved/fixed patch I have locally is:

```patch
--- a/sphinx/extension.py
+++ b/sphinx/extension.py
@@ -51,13 +51,9 @@ def verify_needs_extensions(app: Sphinx, config: Config) -> None:
     for extname, reqversion in config.needs_extensions.items():
         extension = app.extensions.get(extname)
         if extension is None:
-            logger.warning(
-                __(
-                    'The %s extension is required by needs_extensions settings, '
-                    'but it is not loaded.'
-                ),
-                extname,
-            )
+            msg = __('The %s extension is required by needs_extensions settings, '
+                     'but it is not loaded.')
+            logger.warning(msg, extname)
             continue
 
         fulfilled = True
```

Reply:
That's weird. That's definitely not the right formatting

Reply:
> I'm using ruff 0.3.2.

It was moved up to 0.3.3 a few commits ago, maybe also check that

Reply:
> > I'm using ruff 0.3.2.
> 
> It was moved up to 0.3.3 a few commits ago, maybe also check that

Good catch, thanks!  Same result, remains stable:

```sh
$ ruff check --verbose sphinx/extension.py
[2024-03-19][21:21:17][ruff::resolve][DEBUG] Using configuration file (via parent) at: ~/sphinx/.ruff.toml
[2024-03-19][21:21:17][ruff::commands::check][DEBUG] Identified files to lint in: 5.069557ms
[2024-03-19][21:21:17][ruff::commands::check][DEBUG] Checked 1 files in: 95.619µs
All checks passed!
```

(cleared `.ruff_cache` before the check, too)

Reply:
@jayaddison  see https://github.com/sphinx-doc/sphinx/pull/12146/files#r1532102802

Reply:
That is extremely strange that it passes on my machine.  Ok, ignore my suggestion - I'll spend some time to figure out how that happened soon.  Apologies for the distraction!

Reply:
(and thanks for providing a CI-based confirmation, I should have thought to try that!)

Reply:
no worries cheers

Reply:
Ok; I'm pausing for a moment just in case this is relevant.

When using the full file path (`ruff check sphinx/extension.py`) it seems that our `.ruff.toml` config is ignored.

That's why this passed when I ran the command above.

When I run `ruff check` in the root `sphinx.git` directory with my suggestion in place, it fails (`F821`).

I think that means that something in our project configuration (either `flake8` or `ruff` related) introduces this failure.  I'd like to confirm what that is.

Reply:
Ok, confirmed: it is due to our config choice here: https://github.com/sphinx-doc/sphinx/blob/392358d4de91c560a833074ddf9a4c6fc2b07a62/.ruff.toml#L116-L117